library/cgroups.go

```diff
@@ -0,0 +1,105 @@
+package library
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// CgroupImpl breaks out the cgroup functionality so the tests don't need elevated privileges.
+type CgroupImpl interface {
+	MakeCgroup(cg *ControlGroup, ppid Process) error
+	Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error)
+}
+
+type RealCgroup struct{}
+
+func (r *RealCgroup) MakeCgroup(cg *ControlGroup, ppid Process) error {
+	// makes cgroups in the v2 style of /sys/fs/cgroup/<ppid>
+	group := fmt.Sprintf("/sys/fs/cgroup/%s", ppid)
+	if err := os.MkdirAll(group, 0644); err != nil {
+		if errors.Is(err, fs.ErrExist) {
+			log.Println("cgroup", group, "already exists")
+		} else {
+			log.Println("failed to create cgroup for", group, ":", err)
+			return err
+		}
+	}
+
+	// Now we can write the actual cgroup limits ...
+	for _, l := range cg.Limits {
+		lf := fmt.Sprintf("%s/%s", group, l.Var)
+		setVar(l.Value, lf)
+	}
+	return nil
+}
+
+func (r *RealCgroup) Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error) {
```

Comment:
```suggestion
func (r *RealCgroup) Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(Process), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error) {
```

If you make cleanup a func that takes the arguments it needs, you can probably get rid of the awkward `func cleanup() func() { ... }` wrapper that just passes the process into your `internalCleanup`.

Additionally, this is quite a lot of arguments for a single method. Can you think of a cleaner API that might be easier to use?

Reply:
not sure I follow the `cleanup()` comment. The required parameters are not understood by `CGroupImpl` and seem odd to pass in. Or perhaps I'm missing the point of the comment :)